source/multigrid/mg_transfer_prebuilt.cc

```diff
@@ -280,6 +280,137 @@ namespace
+
+    // localize the copy indices for faster access. Since all access will be
+    // through the ghosted vector in 'data', we can use this (much faster)
+    // option
+    data.copy_indices.resize(dof_handler.get_triangulation().n_global_levels());
+    data.copy_indices_level_mine.resize(dof_handler.get_triangulation().n_global_levels());
+    data.copy_indices_global_mine.resize(dof_handler.get_triangulation().n_global_levels());
+    for (unsigned int level=0; level<dof_handler.get_triangulation().n_global_levels(); ++level)
+      {
+        const Utilities::MPI::Partitioner &global_partitioner =
+          *data.ghosted_global_vector.get_partitioner();
+        const Utilities::MPI::Partitioner &level_partitioner =
+          *data.ghosted_level_vector[level].get_partitioner();
+        // owned-owned case: the locally owned indices are going to control
+        // the local index
+        data.copy_indices[level].resize(copy_indices[level].size());
+        for (unsigned int i=0; i<copy_indices[level].size(); ++i)
+          data.copy_indices[level][i] =
+            std::pair<unsigned int,unsigned int>
+            (global_partitioner.global_to_local(copy_indices[level][i].first),
+             level_partitioner.global_to_local(copy_indices[level][i].second));
+
+        // remote-owned case: the locally owned indices for the level and the
+        // ghost dofs for the global indices set the local index
+        data.copy_indices_level_mine[level].
+        resize(copy_indices_level_mine[level].size());
+        for (unsigned int i=0; i<copy_indices_level_mine[level].size(); ++i)
+          data.copy_indices_level_mine[level][i] =
+            std::pair<unsigned int,unsigned int>
+            (global_partitioner.global_to_local(copy_indices_level_mine[level][i].first),
+             level_partitioner.global_to_local(copy_indices_level_mine[level][i].second));
+
+        // owned-remote case: the locally owned indices for the global dofs
+        // and the ghost dofs for the level indices set the local index
+        data.copy_indices_global_mine[level].
+        resize(copy_indices_global_mine[level].size());
+        for (unsigned int i=0; i<copy_indices_global_mine[level].size(); ++i)
+          data.copy_indices_global_mine[level][i] =
+            std::pair<unsigned int,unsigned int>
+            (global_partitioner.global_to_local(copy_indices_global_mine[level][i].first),
+             level_partitioner.global_to_local(copy_indices_global_mine[level][i].second));
+      }
+    data.perform_plain_copy = data.copy_indices.back().size()
+                              == dof_handler.locally_owned_dofs().n_elements();
+    if (data.perform_plain_copy)
+      {
+        AssertDimension(data.copy_indices_global_mine.back().size(), 0);
+        AssertDimension(data.copy_indices_level_mine.back().size(), 0);
+        for (unsigned int i=0; i<data.copy_indices.back().size(); ++i)
+          if (data.copy_indices.back()[i].first !=
```

Comment:
this catches the case where global DoFs are renumbered, right? Can you leave a comment please?